### DISCUSSION TITLE
How to limit the table to page width?

### DISCUSSION THREAD

**Discussion by vincent7f** [Witiko/markdown#115] [Q&A] [answered]
I tried to create a table with 4 columns and converted to a PDF file, but the 4th column is out of the page. How could I fix it?

My code:
```
| Name | Protect who/what? | Who carries? | Who is involved? |
| :--- | :--- | :--- | :--- |
| Criminal law | society | Law enforcement of federal and state governments | |
| Civial law | business | court | affected parties in business | 
| Administrative law | business of government agencies | federal agency | |
```

![image](https://user-images.githubusercontent.com/45114756/147189334-99961176-0584-4184-8b58-dcedc4a6212d.png)

**Comment by Witiko** [comment]
@vincent7f In LaTeX, you can increase the paper width, increase the text width by decreasing the page margins, decrease the font size, or shorten the texts in the cells.

**Comment by Witiko** [comment]
Furthermore, and that is perhaps the most elegant option, you could make some of the cells into fixed-width paragraphs and these could then break across several lines. However, in these cases, it is often best to escape from Markdown and produce the more difficult tables in LaTeX.

**Comment by Witiko** [comment]
If you can provide a minimum working example including the LaTeX preamble, I can show you how some of the above could be implemented in your document and how they would look.

**Comment by vincent7f** [comment]
Dear @Witiko,

Thanks for your quick response and tips. I tried your suggestion but failed. Actually, I used tabular in Latex to create a table with limited width. But Markdown is my favorite writing style, I hope to resolve this table width problem in Markdown. Then I could keep my writing in Latex with Markdown. 

Here is my sample. Regarding the .tex file couldn't be uploaded, so I changed it to be a .txt file.

[test.tex.txt](https://github.com/Witiko/markdown/files/7772022/test.tex.txt)



> If you can provide a minimum working example including the LaTeX preamble, I can show you how some of the above could be implemented in your document and how they would look.

**Comment by Witiko** [comment]
You can make the last three columns¹ into 2.75cm-wide paragraphs with line breaking.

 ![1i][]

[1i]: https://user-images.githubusercontent.com/603082/147394898-ff5a46fa-087f-4162-9ffb-9986ea6654ce.png
[1]: https://github.com/Witiko/markdown/files/7777707/test.tex.txt

See [`test.tex`][1] for the source code. The limitation is that you would need to adjust the 2.75cm bit before every single table.

***

¹ More specifically, this will apply to all columns without the alignment specified (`| ---- |`).

**Comment by vincent7f** [comment]
Dear Witiko,

I can't find the different between yours and mine. Seems to be the same file. Would you mind to have a check? And could you tell me how to add the line breaking inside the Markdown? I believe the line break would be a better and simple solution for me.

**Comment by Witiko** [comment]
Dear @vincent7f,

> I can't find the different between yours and mine. Seems to be the same file.

I am sorry, my mistake. I updated the source code.

> And could you tell me how to add the line breaking inside the Markdown? I believe the line break would be a better and simple solution for me.

There is currently no such option in the `pipeTables` syntax extension, which is intended for tables with simple single-line cells. Your best bet for tables with multiline cells would be to [use LaTeX](https://tex.stackexchange.com/q/2441/70941).

**Comment by vincent7f** [comment]
> 
Dear @Witiko,

It found that tabularx could support line wrapping inside the cell. The example is as below. I think it's an elegant way to convert the Markdown table to a Latex tabularx table, because it does all the things for the writer. May I seek your help to replace the existing table with tabularx ? Thank you so much.

Example: tabularx 
[tabularx.tex.txt](https://github.com/Witiko/markdown/files/7782165/tabularx.tex.txt)

Screenshot:
![image](https://user-images.githubusercontent.com/45114756/147516683-b15acec6-4ee9-427b-bf73-96e62d4760f3.png)

**Comment by Witiko** [comment, marked as answer]
@vincent7f

> It found that tabularx could support line wrapping inside the cell.

There is a number of readability issues with this solution. For one, the justified text creates large spaces between words. We can easily fix this by replacing justified alignment with ragged-right alignment:

``` tex
\usepackage{ragged2e}
% ...
\begin{tabularx}{\linewidth}{|l|X<{\RaggedRight}|X<{\RaggedRight}|X<{\RaggedRight}|}
% ...
```

Another potential issue is that the horizontal space is divided uniformly between the `X` columns, whereas you may want to make some of the columns wider and others more narrow. That is where the `p{2.75cm}` columns may be more useful than the `X` columns in LaTeX. In the Markdown package, there is currently no way to communicate the desired column width.¹

> May I seek your help to replace the existing table with `tabularx`?

Sure. It is a matter of taking [the example table renderer from the Markdown package][1] and modifying it a little here and there for `tabularx`:

![Untitled_waifu2x_art_scale_tta_1](https://user-images.githubusercontent.com/603082/147709346-95f6f840-99cd-4987-ac0e-fd779114ffde.png "Section 3.3.4.4 of the technical documentation of the Markdown package")

To keep things clean, I suggest you put the code into a separate LaTeX theme file named [`markdownthemevincent7f_tabularx.sty`][2]. Then, you can just type away your document:

``` tex
\documentclass[a4paper, 12pt]{book}

\usepackage[theme = vincent7f/tabularx]{markdown}

\begin{document}

\begin{markdown}

| Name | Protect who/what? | Who carries? | Who is involved? |
| :--- | ---- | ---- | ---- |
| Criminal law | society | Law enforcement of federal and state governments | |
| Civial law | business | court | affected parties  in business | 
| Administrative law | day-to-day business of government agencies | federal agency | |

: Comparison of Criminal law, Civil law and administrative law

\end{markdown}

\end{document}
```

![example](https://user-images.githubusercontent.com/603082/147708529-59ad127d-820d-48c8-932a-3ea70abe2444.png "How things will look")

 [1]: https://mirrors.ctan.org/macros/generic/markdown/markdown.pdf#page=183 (Section 3.3.4.4 of the technical documentation of the Markdown package)
 [2]: https://github.com/Witiko/markdown/files/7790959/markdownthemevincent7f_tabularx.sty.txt

***

¹ Another software package, Pandoc, allows you to specify the width of a column with the number of dashes below the header (for example, `| --- | - | - | - |` says that the first column gets half the width). There is no such support in Markdown at the moment, because it is not clear to me whether this is more benetifial than it is confusing to the end user. I may rethink this in the future.

**Reply by vincent7f** [reply]
Dear @Witiko,

Thank you for your patience and effort. It's very helpful for me. I could create my own markdown package based on your document.

**Reply by Witiko** [reply]
@vincent7f Glad I could help.
